src/affexpr.jl

```diff
@@ -15,23 +15,84 @@
 # Operator overloads in src/operators.jl
 #############################################################################
 
+# Utilities for OrderedDict
+function add_or_set!(dict::OrderedDict{K,V}, k::K, v::V) where {K,V}
+    # TODO: This unnecessarily requires two lookups for k.
+    # TODO: Decide if we want to drop zeros here after understanding the
+    # performance implications.
+    dict[k] = get!(dict, k, zero(V)) + v
+    return dict
+end
+
+function new_ordered_dict(::Type{K}, ::Type{V}, kv::AbstractArray{<:Pair}) where {K,V}
+    dict = OrderedDict{K,V}()
+    sizehint!(dict, length(kv))
+    for pair in kv
+        add_or_set!(dict, convert(K, pair.first), convert(V, pair.second))
+    end
+    return dict
+end
+
+function new_ordered_dict(::Type{K}, ::Type{V}, kv::Pair...) where {K,V}
+    dict = OrderedDict{K,V}()
+    sizehint!(dict, length(kv))
+    for pair in kv
+        add_or_set!(dict, convert(K, pair.first), convert(V, pair.second))
+    end
+    return dict
+end
+
+
+
 #############################################################################
 # GenericAffExpr
 # ∑ aᵢ xᵢ  +  c
 mutable struct GenericAffExpr{CoefType,VarType} <: AbstractJuMPScalar
-    vars::Vector{VarType}
-    coeffs::Vector{CoefType}
     constant::CoefType
+    terms::OrderedDict{VarType,CoefType}
 end
 
 variablereftype(::GenericAffExpr{C, V}) where {C, V} = V
 
-Base.iszero(a::GenericAffExpr) = isempty(a.vars) && iszero(a.constant)
-Base.zero(::Type{GenericAffExpr{C,V}}) where {C,V} = GenericAffExpr{C,V}(V[],C[],zero(C))
-Base.one(::Type{GenericAffExpr{C,V}}) where { C,V} = GenericAffExpr{C,V}(V[],C[], one(C))
+function GenericAffExpr(constant::V, kv::AbstractArray{Pair{K,V}}) where {K,V}
+    return GenericAffExpr{V,K}(constant, new_ordered_dict(K, V, kv))
+end
+
+function GenericAffExpr(constant::V, kv::Pair{K,V}...) where {K,V}
+    return GenericAffExpr{V,K}(constant, new_ordered_dict(K, V, kv...))
+end
+
+function GenericAffExpr{V,K}(constant, kv::AbstractArray{<:Pair}) where {K,V}
+    return GenericAffExpr{V,K}(convert(V, constant), new_ordered_dict(K, V, kv))
+end
+
+function GenericAffExpr{V,K}(constant, kv::Pair...) where {K,V}
+    return GenericAffExpr{V,K}(convert(V, constant), new_ordered_dict(K, V, kv...))
+end
+
+Base.iszero(a::GenericAffExpr) = isempty(a.terms) && iszero(a.constant)
+Base.zero(::Type{GenericAffExpr{C,V}}) where {C,V} = GenericAffExpr{C,V}(zero(C), OrderedDict{V,C}())
+Base.one(::Type{GenericAffExpr{C,V}}) where {C,V}  = GenericAffExpr{C,V}(one(C), OrderedDict{V,C}())
 Base.zero(a::GenericAffExpr) = zero(typeof(a))
 Base.one( a::GenericAffExpr) =  one(typeof(a))
-Base.copy(a::GenericAffExpr) = GenericAffExpr(copy(a.vars),copy(a.coeffs),copy(a.constant))
+Base.copy(a::GenericAffExpr) = GenericAffExpr(copy(a.constant), copy(a.terms))
+
+GenericAffExpr{C, V}() where {C, V} = zero(GenericAffExpr{C, V})
+
+function map_coefficients!(a::GenericAffExpr, f::Function)
```

Comment:
It would be more consistent with `Base.map` to have `f` as a first parameter

Reply:
Good point, done. I renamed `map_coefficients!` to `map_coefficients_inplace!`, because `map!` is not an in-place operation (it only stores the output in a given container).